app/services/user_locale_setter.rb

```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+class UserLocaleSetter
+  def initialize(current_user, params_locale, cookies)
+    @current_user = current_user
+    @params_locale = params_locale
+    @cookies = cookies
+  end
+
+  def set_locale
+    save_locale_from_params
+
+    I18n.locale = valid_current_locale
+  end
+
+  def ensure_valid_locale_persisted
+    return unless current_user && !available_locale?(current_user.locale)
+
+    current_user.update!(locale: valid_current_locale)
+  end
+
+  def self.valid_locale_for_user(user)
+    if user.andand.locale.present? && available_locale?(user.locale)
+      user.locale
+    else
+      I18n.default_locale
+    end
+  end
+
+  def self.available_locale?(locale)
+    Rails.application.config.i18n.available_locales.include?(locale)
+  end
+
+  private
+
+  attr_reader :current_user, :params_locale, :cookies
+
+  def save_locale_from_params
+    return unless params_locale && available_locale?(params_locale)
+
+    current_user&.update!(locale: params_locale)
+    cookies[:locale] = params_locale
+  end
+
+  def available_locale?(locale)
+    self.class.available_locale?(locale)
+  end
+
+  def current_user_locale
```

Comment:
this is only used in valid_current_locale, in my book it should be after valid_current_locale. I wonder if you follow the opposite positioning rule, it's possible and equally valid :-)